mono/metadata/w32socket-unix.c

```diff
@@ -320,11 +327,58 @@ mono_w32socket_recvfrom (SOCKET sock, char *buf, int len, int flags, struct sock
 
 	info = mono_thread_info_current ();
 
+	if (blocking) {
+		memset(&tstart, 0, timevallen);
+		memset(&tlimit, 0, timevallen);
+		memset(&timeout, 0, timevallen);
+
+		MONO_ENTER_GC_SAFE;
+		getsockopt (((MonoFDHandle*) sockethandle)->fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, &timevallen);
```

Comment:
Nitpick: introduce `MonoFDHandle *handle = (MonoFDHandle *) sockethandle` earlier instead of repeating this cast.

This also adds at least one extra syscall to every blocking `recvfrom` call. This is probably not a big deal. Another alternative, which I'm noting here for posterity and not because I think you should implement it right now, is to unconditonally poll for the current time (which goes through a vdso on Linux and FreeBSD) at the beginning of a blocking `recvfrom`; if you do fail with an `EINTR`, then you grab the current value of `SO_RCVTIMEO` and adjust the socket timeout on each loop.